### ISSUE TITLE
Problems submitting transactions

### ISSUE THREAD

**Issue by aspect** [kaspanet/kaspad#1600]
Over the past few days, while performing various transaction testing we've encountered faults that look as follows:

```
getting error API error (100): Rejected transaction de7e1f104a7d62e5a1db16c3086730bf531858a48be017cd771de4dfba1f21dd:
output (acd2ce956da06933a234a9eb47459554d911764f365b5ab23c175ef75675c85c: 0) already spent by transaction
57837cd3bcef516d6c77ea1d2fb20a1bf9020b0860ac303f9607451d046fa92b in the memory pool
```

```
2021-03-11 03:37:31 [KaspaWallet] error: ApiError: API error (100):
Rejected transaction e46a5c329c5af6efdffecaa47fb3d781ab3c238af50f272974a9df811f975b8f:
already have transaction e46a5c329c5af6efdffecaa47fb3d781ab3c238af50f272974a9df811f975b8f
```

There is nothing specific about these errors in terms of being able to replicate them.  
However, don't know if this is related, but we did encounter an issue that
we are able to replicate, hopefully, it will shed some light on the above errors as well.

The problem came up when we implemented the "compound" functionality in the wallet. The compound function takes all available UTXOs and creates a single transaction using all available UTXOs as inputs with a single output to your own address.  (The functionality is intended to offer the user to compound the wallet if it has too many UTXOs)

A compound transaction is "fee-inclusive" and uses wallet API for fee estimates, then reducing the total output amount, less fees.

Here is what takes place:

1) The wallet goes through "discovery" phase, enumerating UTXOs that belong to the wallet address list.
2) All UTXOs are merged into a single 'compound' transactions
3) Transaction is posted to Kaspad with success

At this point, we expect that upon accepting the transaction, we will receive
**UtxoChanged** notification removing UTXOs used in this transaction, however, this does not occur.

There is no notification and it seems the transaction gets stuck - tx is accepted, but its UTXOs are not removed from utxoindex but you can not reuse the same UTXOs as a part of another transaction as internally Kaspad sees these UTXOs as spent.

To replicate you can use the latest CLI wallet and do the following series of operations:

Install CLI:
`npm install -g @kaspa/wallet-cli@1.1.0`
https://www.npmjs.com/package/@kaspa/wallet-cli
or direct from git:
`npm install -g aspectron/kaspa-wallet-cli`

(1.1.0 is the latest version that includes the compound feature)

If you have not run CLI or KDX previously, you need to make sure kaspad is running and you can create a wallet using  `kaspa-wallet --testnet create`.  (**NOTE:** you can add `--force` flag to create a seed that is not encrypted or later run `kaspa-wallet permanently-decrypt` to decrypt your seed - this will cause Kaspa Wallet to no longer ask you for the wallet password - useful only during the development and testing).

Get the address of the wallet:
`kaspa-wallet --testnet address`

Mine 1,000+ UTXOs to the addresses
While mining you can run `kaspa-wallet --testnet monitor`

Once the mining is done, check the balance
`kaspa-wallet --testnet balance`
(you will see ~1000 UTXOs available and the calculated balance)

Create compound transaction
`kaspa-wallet --testnet compound`

Check the balance again - it should be 0 UTXOs and 0 KAS.

You can use `kaspa-wallet --testnet info` to get a more detailed information snapshot about the wallet.

Example:

```

$ node kaspa-wallet --testnet info
...
current address: kaspatest:qq5nca0jufeku4mn6ecv50spa3ydfw686ulfal0a56
balance available: 108.54837527 KAS  pending: 0 KAS  total: 108.54837527 KAS
receive addresses used: 2
change addresses used:  20
UTXO storage:
  address: kaspatest:qpvywd4az7zvesze8w3ahafcs55tjks7ncy7adlg7e  UTXOs: 1113
  address: kaspatest:qzcul8vj9cluv532lwlvapq6qn5gnnn645rnkv7taf  UTXOs: 1
  address: kaspatest:qq26fex5mfslmqpt3lyyuzuu6w7e9m3xvyh205czah  UTXOs: 1
  address: kaspatest:qz8fs0zslrcuu536vncltmpmkd9w7sszjcvg0grew3  UTXOs: 1
...

$node kaspa-wallet --testnet compound
...
2021-03-11 03:34:41 [KaspaWallet] info: tx ... required data fee: 0.00168653 (1116 UTXOs)
2021-03-11 03:34:41 [KaspaWallet] info: tx ... resulting total: 108.54837527
...
2021-03-11 03:35:14 [KaspaWallet] info: compound transactions response {
  txid: 'e46a5c329c5af6efdffecaa47fb3d781ab3c238af50f272974a9df811f975b8f'
}
...

$ node kaspa-wallet --testnet info
(this should now be 0s across the board and include only a single UTXO from the compound tx)
...
current address: kaspatest:qq5nca0jufeku4mn6ecv50spa3ydfw686ulfal0a56
balance available: 108.54837527 KAS  pending: 0 KAS  total: 108.54837527 KAS
receive addresses used: 2
change addresses used:  20
UTXO storage:
  address: kaspatest:qpvywd4az7zvesze8w3ahafcs55tjks7ncy7adlg7e  UTXOs: 1113
  address: kaspatest:qzcul8vj9cluv532lwlvapq6qn5gnnn645rnkv7taf  UTXOs: 1
  address: kaspatest:qq26fex5mfslmqpt3lyyuzuu6w7e9m3xvyh205czah  UTXOs: 1
  address: kaspatest:qz8fs0zslrcuu536vncltmpmkd9w7sszjcvg0grew3  UTXOs: 1
...

$node kaspa-wallet --testnet compound
...
2021-03-11 03:37:31 [KaspaWallet] error: ApiError: API error (100): 
Rejected transaction e46a5c329c5af6efdffecaa47fb3d781ab3c238af50f272974a9df811f975b8f: 
already have transaction e46a5c329c5af6efdffecaa47fb3d781ab3c238af50f272974a9df811f975b8f
```

It looks to be some sort of an edge case that is agitated by using up all UTXOs.  As I mentioned, the "UTXO already used" errors do show up at random so the above-described issue may not be related.

**Comment by nanxin0366** [comment]
We are currently facing this issue. Has it been resolved now?

**Comment by aspect** [comment]
This specific issue has been resolved (it's very old). However, such issues can occur under some extremely rare conditions. Usually the node will eventually resolve them. To manually tackle them, you should try resyncing the node by deleting its database and restarting it, or you can also try rusty Kaspa (it has much faster sync times).

**Comment by aspect** [comment]
I just discussed potential issues with other developers, please switch to Rusty Kaspa.

**Comment by ash10121** [comment]
hello i was wondering if you could help with this issue 
or if you could just shed a bit of light 
we are currently trying to implement our coin into the web wallet and we get this error 
our coin is using pyrinhash blake3


![image](https://github.com/kaspanet/kaspad/assets/75103366/f0f9a892-82e9-46f6-824f-6b0b6ccb4688)

**Comment by nanxin0366** [comment]
Yes, we have resolved it. We will send the next transaction after the
previous one has been included in a block.

ash10121 ***@***.***> 于2024年3月23日周六 11:02写道：

**Comment by shinobione** [comment]
Hello.
I have a related problem i guess, as a simple user of kaspa coins.
I sent some of my kaspa coins from kaspa wallet to an exchange. here is the transaction link : [https://explorer.kaspa.org/txs/304aee60b2ea52442c8d107dad1ec287d801e11b7cdd35871ff21ff3492cf917](https://explorer.kaspa.org/txs/304aee60b2ea52442c8d107dad1ec287d801e11b7cdd35871ff21ff3492cf917)
However, since I confirmed the transaction, it's like stuck in the mempool, and my wallet shows 0 KAS, nothing on the exchange, 

edit : now it shows my KAS amount back, like I didnt make any transaction, but when i try to move coins, I have an error like OP : 

> Error
> ×
> API error (100): Rejected transaction add9cd863911fe2be059cd4e05c4547f22d88a68191e7ee933ba72d651685e81: output (0706a1d0417239a96d39a84013bbc99b0021b56ef6d5a6a4262b43c3389e2d0c, 1) already spent by transaction 304aee60b2ea52442c8d107dad1ec287d801e11b7cdd35871ff21ff3492cf917 in the mempool
> 

I somewhat rely on this and i need to get some USDT from those KASPA. but i'm stuck...
i don't get it.

**Comment by itxp2008** [comment]
> Hello. I have a related problem i guess, as a simple user of kaspa coins. I sent some of my kaspa coins from kaspa wallet to an exchange. here is the transaction link : https://explorer.kaspa.org/txs/304aee60b2ea52442c8d107dad1ec287d801e11b7cdd35871ff21ff3492cf917 However, since I confirmed the transaction, it's like stuck in the mempool, and my wallet shows 0 KAS, nothing on the exchange,
> 
> edit : now it shows my KAS amount back, like I didnt make any transaction, but when i try to move coins, I have an error like OP :
> 
> > Error
> > ×
> > API error (100): Rejected transaction add9cd863911fe2be059cd4e05c4547f22d88a68191e7ee933ba72d651685e81: output (0706a1d0417239a96d39a84013bbc99b0021b56ef6d5a6a4262b43c3389e2d0c, 1) already spent by transaction 304aee60b2ea52442c8d107dad1ec287d801e11b7cdd35871ff21ff3492cf917 in the mempool
> 
> I somewhat rely on this and i need to get some USDT from those KASPA. but i'm stuck... i don't get it.

Have you managed to fix it? I have the same issue here
